src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -78,6 +78,122 @@
+    that the search through ancestor nodes stops when the scope_limit node
+    is reached i.e. ancestors of the scope_limit node are not searched.
+
+    :param location: PSyIR node from which to operate.
+    :type location: :py:class:`psyclone.psyir.nodes.Node`
+    :param str name: the name of the symbol.
+    :param scope_limit: optional Node which limits the symbol \
+        search space to the symbol tables of the nodes within the \
+        given scope. If it is None (the default), the whole \
+        scope (all symbol tables in ancestor nodes) is searched \
+        otherwise ancestors of the scope_limit node are not \
+        searched.
+    :type scope_limit: :py:class:`psyclone.psyir.nodes.Node` or \
+        `NoneType`
+
+    :returns: the matching symbol.
+    :rtype: :py:class:`psyclone.psyir.symbols.Symbol`
+
+    :raises TypeError: if the supplied scope_limit is not a Node.
+    :raises ValueError: if the supplied scope_limit node is not an \
+        ancestor of the supplied node.
+    :raises SymbolError: if no matching symbol is found and there are \
+        no ContainerSymbols from which it might be brought into scope.
+
+    '''
+    if not isinstance(location, Node):
+        raise TypeError(
+            "The location argument '{0}' provided to _find_or_create_imported"
+            "_symbol() is not of type `Node`.".format(str(location)))
+
+    if scope_limit:
+        # Validate the supplied scope_limit
+        if not isinstance(scope_limit, Node):
+            raise TypeError(
+                "The scope_limit argument '{0}' provided to _find_or_"
+                "create_imported_symbol() is not of type `Node`."
+                "".format(str(scope_limit)))
+
+        # Check that the scope_limit Node is an ancestor of this
+        # Reference Node and raise an exception if not.
+        mynode = location.parent
+        while mynode is not None:
+            if mynode is scope_limit:
+                # The scope_limit node is an ancestor of the
+                # supplied node.
+                break
+            mynode = mynode.parent
+        else:
+            # The scope_limit node is not an ancestor of the
+            # supplied node so raise an exception.
+            raise ValueError(
+                "The scope_limit node '{0}' provided to _find_or_create"
+                "_imported_symbol() is not an ancestor of this "
+                "node '{1}'.".format(str(scope_limit), str(location)))
+
+    # Keep a reference to the most local SymbolTable with a wildcard
+    # import in case we need to create a Symbol.
+    first_symbol_table = None
+    test_node = location
+
+    # Iterate over ancestor Nodes of this Node.
+    while test_node:
+        # For simplicity, test every Node for the existence of a
+        # SymbolTable (rather than checking for the particular
+        # Node types which we know to have SymbolTables).
+        if hasattr(test_node, 'symbol_table'):
```

Comment:
I tried but unfortunately it's not easy because there is the break condition that `test_node is scope_limit` and scope_limit can be any node (not just scoping nodes). Therefore the breaking condition would have to check that descendants aren't the scope limit. So for now I leave it as it is.

Ultimately I would like to see if this whole functionality can be integrated into the Symbol Table, essentially what this does different is that if a `interface=UnresolvedInterface()` is given, it either exist or the symbol_table should have a wildcard import container, otherwise fail. But I leave this for the future.